### PR TITLE
Make the code-gen part of macro-expansion deterministic

### PR DESCRIPTION
**Pull request by regexident**
The current codegen implementation in `ascent_macro` is non-deterministic due to the use of `HashSet` and `HashMap`.

This PR sorts the set/map-based items before emitting code for them, resulting in reproducible expansions.

One of the motivations (other than making builds reproducible) is being able to introduce a suite of snapshot tests to the project (https://github.com/s-arash/ascent/pull/35), which greatly helps catching regressions in proc-macros.

### CODE REVIEW

**Comment by regexident** [comment]
@s-arash I've rebased the PR onto `main`, resolving the conflicts.

File: ascent_macro/src/ascent_codegen.rs [hunk 1]
```diff
@@ -662,7 +675,8 @@ fn compile_update_indices_function_body(mir: &AscentMir) -> proc_macro2::TokenSt
       };
 
       let mut update_indices = vec![];
-      for ind in indices_set.iter(){
+      let sorted_indices = indices_set.iter().sorted_by_cached_key(|rel| rel.ir_name());
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:678]: Let's inline `sorted_indices` here

File: ascent_macro/src/ascent_codegen.rs [hunk 2]
```diff
@@ -992,7 +1006,8 @@ fn head_clauses_structs_and_update_code(rule: &MirRule, scc: &MirScc, mir: &Asce
       let new_ref = if !mir.is_parallel { quote!{&mut} } else { quote!{&} };
       let mut used_fields = HashSet::new();
       if let Some(rel_indices) = rel_indices {
-         for rel_ind in rel_indices.iter(){
+         let sorted_rel_indices = rel_indices.iter().sorted_by_cached_key(|rel| rel.ir_name());
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:1009]: Let's inline `sorted_rel_indices` here.

File: ascent_macro/src/ascent_codegen.rs [hunk 3]
```diff
@@ -583,7 +594,8 @@ fn compile_mir_scc(mir: &AscentMir, scc_ind: usize) -> proc_macro2::TokenStream
 
 fn compile_relation_sizes_body(mir: &AscentMir) -> proc_macro2::TokenStream {
    let mut write_sizes = vec![];
-   for r in mir.relations_ir_relations.keys().sorted_by_key(|r| &r.name) {
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:586]: Let's undo the change here.

File: ascent_macro/src/ascent_codegen.rs [hunk 4]
```diff
@@ -89,8 +91,8 @@ pub(crate) fn compile_mir(mir: &AscentMir, is_ascent_run: bool) -> proc_macro2::
    let mut field_type_names = HashSet::<String>::new();
    let mut lat_field_type_names = HashSet::<String>::new();
 
-   
-   for relation in mir.relations_ir_relations.keys() {
+   let sorted_relations = mir.relations_ir_relations.keys().sorted_by_key(|rel| &rel.name);
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:94]: Let's inline `sorted_relations` here

File: ascent_macro/src/ascent_codegen.rs [hunk 5]
```diff
@@ -115,7 +117,8 @@ pub(crate) fn compile_mir(mir: &AscentMir, is_ascent_run: bool) -> proc_macro2::
    }
 
    let mut relation_initializations = vec![];
-   for (rel, md) in mir.relations_metadata.iter() {
+   let sorted_relations_metadata = mir.relations_metadata.iter().sorted_by_key(|(rel, _)| &rel.name);
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:120]: Let's inline `sorted_relations_metadata` here.

**Review by s-arash — COMMENTED** [review]
Thanks for the PR. I agree the generated code should be deterministic. Left a few comments, otherwise the PR looks good.

**Comment by regexident** [comment]
> Left a few comments, otherwise the PR looks good.

Fixed!

**Review by s-arash — APPROVED** [review]
